File: vane-core/src/main/java/org/oddlama/vane/core/data/CooldownData.java
package org.oddlama.vane.core.data;

import org.bukkit.NamespacedKey;
import org.bukkit.persistence.PersistentDataContainer;
import org.bukkit.persistence.PersistentDataHolder;
import org.bukkit.persistence.PersistentDataType;

public class CooldownData {

    private final NamespacedKey key;
    private final long cooldownTime;

    public CooldownData(NamespacedKey key, long cooldownTime) {
        this.key = key;
        this.cooldownTime = cooldownTime;
    }

    /**
     * Updates the cooldown data if and only if the cooldown has been exceeded.
     *
     * @return returns true if cooldownTime has been exceeded.
     */
    public boolean checkOrUpdateCooldown(final PersistentDataHolder holder) {
        final var persistentData = holder.getPersistentDataContainer();
        final var lastTime = persistentData.getOrDefault(key, PersistentDataType.LONG, 0L);
        final var now = System.currentTimeMillis();
        if (now - lastTime < cooldownTime) {
            return false;
        }

        persistentData.set(key, PersistentDataType.LONG, now);
        return true;
    }

    /**
     * @return Gets the status of the cooldown without updating
     */
    public boolean peekCooldown(PersistentDataHolder holder) {
        PersistentDataContainer persistentData = holder.getPersistentDataContainer();
        Long lastTime = persistentData.getOrDefault(this.key, PersistentDataType.LONG, 0L);
        long now = System.currentTimeMillis();
        return now - lastTime >= this.cooldownTime;
    }

    public void clear(final PersistentDataHolder holder) {
        final var persistentData = holder.getPersistentDataContainer();
        persistentData.remove(key);
    }
}

File: vane-core/src/main/kotlin/org/oddlama/vane/core/data/CooldownData.kt
package org.oddlama.vane.core.data

import org.bukkit.NamespacedKey
import org.bukkit.persistence.PersistentDataHolder
import org.bukkit.persistence.PersistentDataType

/**
 * Persistent-data based cooldown helper.
 *
 * @param key persistent data key used to store last-use timestamp.
 * @param cooldownTime cooldown duration in milliseconds.
 */
class CooldownData(private val key: NamespacedKey, private val cooldownTime: Long) {
    /** Returns elapsed milliseconds since last recorded usage. */
    private fun PersistentDataHolder.elapsedSinceLast(): Long =
        System.currentTimeMillis() - persistentDataContainer.getOrDefault(key, PersistentDataType.LONG, 0L)

    /**
     * Updates the cooldown data if and only if the cooldown has been exceeded.
     *
     * @return true if cooldownTime has been exceeded.
     */
    fun checkOrUpdateCooldown(holder: PersistentDataHolder): Boolean {
        if (holder.elapsedSinceLast() < cooldownTime) return false
        holder.persistentDataContainer.set(key, PersistentDataType.LONG, System.currentTimeMillis())
        return true
    }

    /**
     * @return the status of the cooldown without updating.
     */
    fun peekCooldown(holder: PersistentDataHolder): Boolean = holder.elapsedSinceLast() >= cooldownTime

    /** Clears stored cooldown timestamp. */
    fun clear(holder: PersistentDataHolder) = holder.persistentDataContainer.remove(key)
}
